torch/csrc/distributed/c10d/reducer.hpp

```diff
@@ -192,10 +184,9 @@ class TORCH_API Reducer {
   void install_futures(c10::List<c10::intrusive_ptr<c10::ivalue::Future>> futs);
 
   // Returns true if we should rebuild buckets, else false. We only rebuild
-  // buckets once after the first iteration and never rebuild them if
-  // find_unused_parameters_.
+  // buckets once after the first iteration.
   inline bool should_rebuild_buckets() const {
-    return (static_graph_ || !find_unused_parameters_) && !has_rebuilt_bucket_;
+    return !has_rebuilt_bucket_;
```

Comment:
Curious, should we add a knob (say, env var) to disable this feature just for debugging purposes?

Reply:
> Do we want to eventually consider rebuilding buckets again when the graph has changed when training dynamic graph? Depending on the use case and how common it is this could also have a perf improvement.
> 
> Also, I'm assuming that this change will cause unused parameters to be allreduced first, because we mark unused parameters immediately when the first hook fires. Is this best for performance or should unused parameters be the last to allreduce?

rebuilt bucket every iteration for dynamic graph will be expensive, as rebuilt bucket needs a broadcast call. I added a comment in the codes.  

in line 836, we pushed the unused parameters at the end of buckets for better performance

Reply:
> Curious, should we add a knob (say, env var) to disable this feature just for debugging purposes?

sounds good to me